Ktlint spotless tasks are always "up-to-date" even when there are changes

I see this issue both with linting Gradle Kotlin DSL files and Kotlin Source files.

Making a change to source code after running spotless and getting a success the task seems to get stuck as always "up-to-date".

A simple example of this is an updated version of the `KotlinExtensionTest::testWithHeader` that I wrote.
The test as it is in master currently passes 🎉 .

However, changing the test so that it looks like this:
```java
@Test
public void testWithHeader() throws IOException {
	write("build.gradle",
			"plugins {",
			"    id 'nebula.kotlin' version '1.0.6'",
			"    id 'com.diffplug.gradle.spotless'",
			"}",
			"repositories { mavenCentral() }",
			"spotless {",
			"    kotlin {",
			"        licenseHeader('" + HEADER + "')",
			"        ktlint()",
			"    }",
			"}");

	// First run, this will put the task into it's "up-to-date" state.
	runTestWithHeader();
	// Second run should see the changes and the task should not be "up-to-date".
	runTestWithHeader();
}

private void runTestWithHeader() throws IOException {
	final File testFile = write("src/main/kotlin/test.kt", getTestResource("kotlin/licenseheader/KotlinCodeWithoutHeader.test"));
	final String original = read(testFile.toPath());
	gradleRunner().withArguments("spotlessApply").build();
	final String result = read(testFile.toPath());
	Assertions
			.assertThat(result)
			// Make sure the header gets added.
			.startsWith(HEADER)
			// Make sure that the rest of the file is still there with nothing removed.
			.endsWith(original)
			// Make sure that no additional stuff got added to the file.
			.contains(HEADER + '\n' + original);
}
```

Results in this test failure:
```
java.lang.AssertionError: 
Expecting:
 <"@file:JvmName("SomeFileName")
package my.test

object AnObject { }

">
to start with:
 <"// License Header">


	at com.diffplug.gradle.spotless.KotlinExtensionTest.runTestWithHeader(KotlinExtensionTest.java:77)
	at com.diffplug.gradle.spotless.KotlinExtensionTest.testWithHeader(KotlinExtensionTest.java:66)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:55)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
```

You will notice that the failure stack trace is from the second call to `runTestWithHeader` meaning that the second run of `spotlessApply` didn't reformat the source code.

--- a/lib/src/main/java/com/diffplug/spotless/PaddedCellBulk.java
+++ b/lib/src/main/java/com/diffplug/spotless/PaddedCellBulk.java
@@ -191,6 +191,11 @@ public final class PaddedCellBulk {
 
 	/** Performs the typical spotlessApply, but with PaddedCell handling of misbehaving FormatterSteps. */
 	public static void apply(Formatter formatter, File file) throws IOException {
+		applyAnyChanged(formatter, file);
+	}
+
+	/** Performs the typical spotlessApply, but with PaddedCell handling of misbehaving FormatterSteps. */
+	public static boolean applyAnyChanged(Formatter formatter, File file) throws IOException {
 		Objects.requireNonNull(formatter, "formatter");
 		Objects.requireNonNull(file, "file");
 
@@ -206,14 +211,14 @@ public final class PaddedCellBulk {
 		// if F(input) == input, then the formatter is well-behaving and the input is clean
 		byte[] formattedBytes = formatted.getBytes(formatter.getEncoding());
 		if (Arrays.equals(rawBytes, formattedBytes)) {
-			return;
+			return false;
 		}
 
 		// F(input) != input, so we'll do a padded check
 		PaddedCell cell = PaddedCell.check(formatter, file, rawUnix);
 		if (!cell.isResolvable()) {
 			// nothing we can do, but check will warn and dump out the divergence path
-			return;
+			return false;
 		}
 
 		// get the canonical bytes
@@ -223,6 +228,9 @@ public final class PaddedCellBulk {
 		if (!Arrays.equals(rawBytes, canonicalBytes)) {
 			// and write them to disk if needed
 			Files.write(file.toPath(), canonicalBytes, StandardOpenOption.TRUNCATE_EXISTING);
+			return true;
+		} else {
+			return false;
 		}
 	}
 

--- a/plugin-gradle/src/main/java/com/diffplug/gradle/spotless/SerializableMisc.java
+++ b/plugin-gradle/src/main/java/com/diffplug/gradle/spotless/SerializableMisc.java
@@ -1,0 +1,68 @@
+/*
+ * Copyright 2016 DiffPlug
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.diffplug.gradle.spotless;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.io.OutputStream;
+import java.io.Serializable;
+
+import com.diffplug.common.base.Errors;
+import com.diffplug.common.base.Preconditions;
+import com.diffplug.common.io.Files;
+
+class SerializableMisc {
+	static void toFile(Serializable obj, File file) {
+		try {
+			java.nio.file.Files.createDirectories(file.getParentFile().toPath());
+			try (OutputStream output = Files.asByteSink(file).openBufferedStream()) {
+				toStream(obj, output);
+			}
+		} catch (IOException e) {
+			throw Errors.asRuntime(e);
+		}
+	}
+
+	static <T extends Serializable> T fromFile(Class<T> clazz, File file) {
+		try (InputStream input = Files.asByteSource(file).openBufferedStream()) {
+			return fromStream(clazz, input);
+		} catch (IOException e) {
+			throw Errors.asRuntime(e);
+		}
+	}
+
+	static void toStream(Serializable obj, OutputStream stream) {
+		try (ObjectOutputStream objectOutput = new ObjectOutputStream(stream)) {
+			objectOutput.writeObject(obj);
+		} catch (IOException e) {
+			throw Errors.asRuntime(e);
+		}
+	}
+
+	@SuppressWarnings("unchecked")
+	static <T> T fromStream(Class<T> clazz, InputStream stream) {
+		try (ObjectInputStream objectInput = new ObjectInputStream(stream)) {
+			T object = (T) objectInput.readObject();
+			Preconditions.checkArgument(clazz.isInstance(object), "Requires class %s, was %s", clazz, object);
+			return object;
+		} catch (ClassNotFoundException | IOException e) {
+			throw Errors.asRuntime(e);
+		}
+	}
+}

--- a/plugin-gradle/src/main/java/com/diffplug/gradle/spotless/SpotlessTask.java
+++ b/plugin-gradle/src/main/java/com/diffplug/gradle/spotless/SpotlessTask.java
@@ -19,6 +19,7 @@ import static com.diffplug.gradle.spotless.PluginGradlePreconditions.requireElem
 
 import java.io.File;
 import java.io.IOException;
+import java.io.Serializable;
 import java.nio.charset.Charset;
 import java.nio.file.Files;
 import java.nio.file.StandardOpenOption;
@@ -32,10 +33,11 @@ import org.gradle.api.DefaultTask;
 import org.gradle.api.GradleException;
 import org.gradle.api.tasks.Input;
 import org.gradle.api.tasks.InputFiles;
-import org.gradle.api.tasks.SkipWhenEmpty;
 import org.gradle.api.tasks.TaskAction;
 import org.gradle.api.tasks.incremental.IncrementalTaskInputs;
 
+import com.diffplug.common.collect.ImmutableList;
+import com.diffplug.common.collect.Iterables;
 import com.diffplug.spotless.FormatExceptionPolicy;
 import com.diffplug.spotless.FormatExceptionPolicyStrict;
 import com.diffplug.spotless.Formatter;
@@ -93,14 +95,24 @@ public class SpotlessTask extends DefaultTask {
 
 	protected Iterable<File> target;
 
-	@InputFiles
-	@SkipWhenEmpty
 	public Iterable<File> getTarget() {
 		return target;
 	}
 
 	public void setTarget(Iterable<File> target) {
 		this.target = requireElementsNonNull(target);
+	}
+
+	/** Internal use only. */
+	@InputFiles
+	@Deprecated
+	public Iterable<File> getInternalTarget() {
+		// used to combine the special cache file and the real target
+		return Iterables.concat(ImmutableList.of(getCacheFile()), target);
+	}
+
+	private File getCacheFile() {
+		return new File(getProject().getBuildDir(), getName());
 	}
 
 	protected List<FormatterStep> steps = new ArrayList<>();
@@ -160,30 +172,64 @@ public class SpotlessTask extends DefaultTask {
 		List<File> outOfDate = new ArrayList<>();
 		inputs.outOfDate(inputDetails -> {
 			File file = inputDetails.getFile();
-			if (file.isFile()) {
+			if (file.isFile() && !file.equals(getCacheFile())) {
 				outOfDate.add(file);
 			}
 		});
+		// load the files that were changed by the last run
+		// because it's possible the user changed them back to their
+		// unformatted form, so we need to treat them as dirty
+		// (see bug #144)
+		if (getCacheFile().exists()) {
+			LastApply lastApply = SerializableMisc.fromFile(LastApply.class, getCacheFile());
+			for (File file : lastApply.changedFiles) {
+				if (!outOfDate.contains(file) && file.exists()) {
+					outOfDate.add(file);
+				}
+			}
+		}
 
 		if (apply) {
-			apply(formatter, outOfDate);
+			List<File> changedFiles = applyAnyChanged(formatter, outOfDate);
+			if (!changedFiles.isEmpty()) {
+				// If any file changed, we need to mark the task as dirty
+				// next time to avoid bug #144.
+				LastApply lastApply = new LastApply();
+				lastApply.timestamp = System.currentTimeMillis();
+				lastApply.changedFiles = changedFiles;
+
+				SerializableMisc.toFile(lastApply, getCacheFile());
+			}
 		}
 		if (check) {
 			check(formatter, outOfDate);
 		}
 	}
 
-	private void apply(Formatter formatter, List<File> outOfDate) throws Exception {
+	static class LastApply implements Serializable {
+		private static final long serialVersionUID = 6245070824310295090L;
+
+		long timestamp;
+		List<File> changedFiles;
+	}
+
+	private List<File> applyAnyChanged(Formatter formatter, List<File> outOfDate) throws Exception {
+		List<File> changed = new ArrayList<>(outOfDate.size());
 		if (isPaddedCell()) {
 			for (File file : outOfDate) {
 				getLogger().debug("Applying format to " + file);
-				PaddedCellBulk.apply(formatter, file);
+				if (PaddedCellBulk.applyAnyChanged(formatter, file)) {
+					changed.add(file);
+				}
 			}
 		} else {
 			boolean anyMisbehave = false;
 			for (File file : outOfDate) {
 				getLogger().debug("Applying format to " + file);
 				String unixResultIfDirty = formatter.applyToAndReturnResultIfDirty(file);
+				if (unixResultIfDirty != null) {
+					changed.add(file);
+				}
 				// because apply will count as up-to-date, it's important
 				// that every call to apply will get a PaddedCell check
 				if (!anyMisbehave && unixResultIfDirty != null) {
@@ -207,6 +253,7 @@ public class SpotlessTask extends DefaultTask {
 				throw PaddedCellGradle.youShouldTurnOnPaddedCell(this);
 			}
 		}
+		return changed;
 	}
 
 	private void check(Formatter formatter, List<File> outOfDate) throws Exception {

--- a/plugin-gradle/src/main/java/com/diffplug/gradle/spotless/SpotlessTask.java
+++ b/plugin-gradle/src/main/java/com/diffplug/gradle/spotless/SpotlessTask.java
@@ -46,9 +46,9 @@ import com.diffplug.spotless.PaddedCellBulk;
 
 public class SpotlessTask extends DefaultTask {
 	// set by SpotlessExtension, but possibly overridden by FormatExtension
-	@Input
 	protected String encoding = "UTF-8";
 
+	@Input
 	public String getEncoding() {
 		return encoding;
 	}
@@ -57,9 +57,9 @@ public class SpotlessTask extends DefaultTask {
 		this.encoding = Objects.requireNonNull(encoding);
 	}
 
-	@Input
 	protected LineEnding.Policy lineEndingsPolicy = LineEnding.UNIX.createPolicy();
 
+	@Input
 	public LineEnding.Policy getLineEndingsPolicy() {
 		return lineEndingsPolicy;
 	}
@@ -69,9 +69,9 @@ public class SpotlessTask extends DefaultTask {
 	}
 
 	// set by FormatExtension
-	@Input
 	protected boolean paddedCell = false;
 
+	@Input
 	public boolean isPaddedCell() {
 		return paddedCell;
 	}
@@ -80,21 +80,21 @@ public class SpotlessTask extends DefaultTask {
 		this.paddedCell = paddedCell;
 	}
 
-	@Input
 	protected FormatExceptionPolicy exceptionPolicy = new FormatExceptionPolicyStrict();
 
 	public void setExceptionPolicy(FormatExceptionPolicy exceptionPolicy) {
 		this.exceptionPolicy = Objects.requireNonNull(exceptionPolicy);
 	}
 
+	@Input
 	public FormatExceptionPolicy getExceptionPolicy() {
 		return exceptionPolicy;
 	}
 
-	@InputFiles
-	@SkipWhenEmpty
 	protected Iterable<File> target;
 
+	@InputFiles
+	@SkipWhenEmpty
 	public Iterable<File> getTarget() {
 		return target;
 	}
@@ -103,9 +103,9 @@ public class SpotlessTask extends DefaultTask {
 		this.target = requireElementsNonNull(target);
 	}
 
-	@Input
 	protected List<FormatterStep> steps = new ArrayList<>();
 
+	@Input
 	public List<FormatterStep> getSteps() {
 		return Collections.unmodifiableList(steps);
 	}

--- a/plugin-gradle/src/test/java/com/diffplug/gradle/spotless/GradleIncrementalResolutionTest.java
+++ b/plugin-gradle/src/test/java/com/diffplug/gradle/spotless/GradleIncrementalResolutionTest.java
@@ -68,11 +68,12 @@ public class GradleIncrementalResolutionTest extends GradleIntegrationTest {
 		// if we change just one file
 		writeState("Abc");
 		// then check runs against just the changed file
-		checkRanAgainst("a");
+		// and also (because #144) the last files to be changed
+		checkRanAgainst("a", "b");
 		// even after failing, still just the one
-		checkRanAgainst("a");
+		checkRanAgainst("a", "b");
 		// and so does apply
-		applyRanAgainst("a");
+		applyRanAgainst("a", "b");
 		applyRanAgainst("a");
 		// until the issue has been fixed
 		applyRanAgainst("");

--- a/plugin-gradle/src/test/java/com/diffplug/gradle/spotless/UpToDateTest.java
+++ b/plugin-gradle/src/test/java/com/diffplug/gradle/spotless/UpToDateTest.java
@@ -1,0 +1,94 @@
+/*
+ * Copyright 2016 DiffPlug
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.diffplug.gradle.spotless;
+
+import java.io.IOException;
+
+import org.junit.Test;
+
+public class UpToDateTest extends GradleIntegrationTest {
+	/** Requires that README be lowercase. */
+	private void writeBuildFile() throws IOException {
+		write("build.gradle",
+				"plugins {",
+				"    id 'com.diffplug.gradle.spotless'",
+				"}",
+				"spotless {",
+				"    format 'misc', {",
+				"        target file('README.md')",
+				"        customLazyGroovy('lowercase') {",
+				"             return { str -> str.toLowerCase(Locale.ROOT) }",
+				"        }",
+				"        bumpThisNumberIfACustomStepChanges(1)",
+				"    }",
+				"}");
+	}
+
+	@Test
+	public void testNormalCase() throws IOException {
+		writeBuildFile();
+		write("README.md", "ABC");
+		// first time, the task runs as expected
+		applyIsUpToDate(false);
+		assertFile("README.md").hasContent("abc");
+		// because a file was changed (by spotless),
+		// up-to-date is false, even though nothing is
+		// going to change during this run.  This second
+		// run is very fast though, because it will
+		// only run on the few files that were changed.
+		applyIsUpToDate(false);
+		// it's not until the third run that everything
+		// is totally up-to-date
+		applyIsUpToDate(true);
+	}
+
+	@Test
+	public void testNearPathologicalCase() throws IOException {
+		writeBuildFile();
+		write("README.md", "ABC");
+		// first time, up-to-date is false
+		applyIsUpToDate(false);
+		assertFile("README.md").hasContent("abc");
+
+		// now we'll change the file
+		write("README.md", "AB");
+		// as expected, the task will run again
+		applyIsUpToDate(false);
+		assertFile("README.md").hasContent("ab");
+		// and it'll take two more runs to get to up-to-date
+		applyIsUpToDate(false);
+		applyIsUpToDate(true);
+	}
+
+	@Test
+	public void testPathologicalCase() throws IOException {
+		writeBuildFile();
+		write("README.md", "ABC");
+		// first time, up-to-date is false
+		applyIsUpToDate(false);
+		assertFile("README.md").hasContent("abc");
+
+		// now we'll change the file back to EXACTLY its original content
+		write("README.md", "ABC");
+		// the task should run again, but instead the next line will
+		// fail an assertion, because the task is actually reported as up-to-date
+		applyIsUpToDate(false);
+		assertFile("README.md").hasContent("abc");
+		// and it'll take two more runs to get to up-to-date
+		applyIsUpToDate(false);
+		applyIsUpToDate(true);
+	}
+}

--- a/testlib/src/main/java/com/diffplug/spotless/ResourceHarness.java
+++ b/testlib/src/main/java/com/diffplug/spotless/ResourceHarness.java
@@ -28,6 +28,8 @@ import java.util.List;
 import java.util.logging.Logger;
 import java.util.stream.Collectors;
 
+import org.assertj.core.api.AbstractFileAssert;
+import org.assertj.core.api.Assertions;
 import org.junit.Assert;
 import org.junit.ComparisonFailure;
 import org.junit.Rule;
@@ -102,6 +104,10 @@ public class ResourceHarness {
 		Files.createDirectories(target.getParent());
 		Files.write(target, content.getBytes(encoding));
 		return target.toFile();
+	}
+
+	protected AbstractFileAssert<?> assertFile(String path) throws IOException {
+		return Assertions.assertThat(newFile(path)).usingCharset(StandardCharsets.UTF_8);
 	}
 
 	protected String read(Path path) throws IOException {

